src/v/bytes/ioarray.h

```diff
@@ -0,0 +1,139 @@
+    // uninitialized.
+    explicit ioarray(size_t size);
+
+    // String-like view of a portion of the ioarray.
+    //
+    // Since strings must be less than 128_KiB (because they are normally
+    // contiguous), this object can hold a view of any chunk of data in the
+    // ioarray that is less than 128_KiB. Since chunks are all 128_KiB (except
+    // the last buffer, which may be smaller), this means that 2 continugous
+    // chunks can represent any string-like view within contents.
+    class string_view {
+    public:
+        string_view() = default;
+        explicit string_view(std::string_view s) { _views[0] = s; }
+        explicit string_view(std::string_view a, std::string_view b)
+          : _views(std::to_array({a, b})) {}
+
+        explicit operator ss::sstring() const;
+        bool operator==(const string_view& other) const = default;
+        auto operator<=>(const string_view& other) const = default;
+        bool operator==(std::string_view other) const;
+        std::strong_ordering operator<=>(std::string_view other) const;
+        size_t size() const { return _views[0].size() + _views[1].size(); }
+        bool empty() const { return _views[0].empty() && _views[1].empty(); }
+        auto as_range() const { return _views | std::views::join; }
+        auto begin() const { return _views.begin(); }
+        auto end() const { return _views.end(); }
+
+    private:
+        std::array<std::string_view, 2> _views;
+    };
+
+    // Create an uninitialized iovector where all the chunks are aligned.
+    // It's assumed that the alignment is a power of two less than
+    // `max_chunk_size` and that size is a multiple of alignment.
+    static ioarray aligned(size_t alignment, size_t size);
+
+    // Create the ioarray from copying out of an iobuf.
+    //
+    // NOTE: This intended to be used for testing.
+    static ioarray copy_from(const iobuf&);
+
+    // Share a portion of ioarray.
+    //
+    // NOTE: This ioarray lengthens the lifetime of the data backing contents,
+    // so it's recommended to use this only for a short time, or to make a copy.
+    //
+    // If this ioarray was allocated aligned, the returned array is no longer
+    // guaranteed to be aligned.
+    ioarray share(size_t pos, size_t length);
+
+    // Convert the ioarray to an iobuf.
+    //
+    // The underlying buffers are shared.
+    iobuf as_iobuf();
+
+    // Get a view of the contents as a string-like object.
+    //
+    // REQUIRES: length <= max_buffer_size
+    string_view read_string(size_t pos, size_t length) const;
+
+    char& operator[](size_t i);
+    char operator[](size_t i) const;
+
+    // Read a little endian fixed32 bit integer from the ioarray.
+    //
+    // REQUIRES: i + sizeof(uint32_t) < size
+    uint32_t read_fixed32(size_t i) const;
+
+    // Remove the last n bytes from this ioarray.
+    void trim_back(size_t n);
+
+    // The size of the ioarray.
+    size_t size() const { return _size; }
```

Comment:
Just wondering if we'll want a memory_usage() call similar to the one in the previous commit (I guess the size may be different for the aligned case?)